### PR TITLE
Add onAuthorizationFailed callback function

### PR DESCRIPTION
**Pull request by mketchum**

### CODE REVIEW

File: demo-server/templates/index-template.handlebars [hunk 1]
```diff
@@ -59,24 +59,32 @@
             initWidget(authnWidget);
         }
 
+        function showResponse(text) {
+          // success callback function
```
Comment (danielhoskin-ping) [inline demo-server/templates/index-template.handlebars:63]: Nit: stale comment?

File: demo-server/templates/index-template.handlebars [hunk 2]
```diff
@@ -59,24 +59,32 @@
...
+
         /**
          * initialize the widget based on the operation mode
          */
         function initWidget(widget) {
           {{#if redirectless}}
           var config = {
-            // client_id: '', // redirectless requires this configuration
-            // response_type: 'token', // redirectless requires this configuration
+            client_id: 'ac_client', // redirectless requires this configuration
```
Comment (danielhoskin-ping) [inline demo-server/templates/index-template.handlebars:80]: Not sure if uncommenting/changing this was only for testing or a new default?

**Review by danielhoskin-ping — APPROVED** [review]

File: demo-server/templates/index-template.handlebars (hunk 1 above)
Comment (mketchum) [reply demo-server/templates/index-template.handlebars]: Good catch, I've removed it

File: demo-server/templates/index-template.handlebars (hunk 2 above)
Comment (mketchum) [reply demo-server/templates/index-template.handlebars]: Can't tell you how many times I told myself I needed to remember to switch this back

**Review by danielhoskin-ping — APPROVED** [review]

File: types/index.d.ts [hunk 3]
```diff
@@ -4,6 +4,7 @@ export interface IRedirectlessConfig {
   client_id?: string,
   response_type?: string,
   onAuthorizationSuccess: (response: Response) => void,
+  onAuthorizationFailed: (response: Response) => void,
```
Comment (khosravian) [inline types/index.d.ts:7]: shouldn't this be optional?

File: src/validators/redirectless.js [hunk 4]
```diff
@@ -13,6 +13,13 @@ const redirectlessConfigValidator = (configuration) => {
     throw new Error('onAuthorizationSuccess attribute must be a function.');
   }
 
+  if (configuration.onAuthorizationFailed) {
+    const onAuthorizationFailed = configuration.onAuthorizationSuccess;
+    if(!validator.isFunction(onAuthorizationFailed)) {
```
Comment (khosravian) [inline src/validators/redirectless.js:18]: missing a space between `if` and `(` 😜

Comment (mketchum) [reply src/validators/redirectless.js]: Good eyes 👀

File: src/utils/redirectless.js [hunk 5]
```diff
@@ -15,6 +15,21 @@ export function completeStateCallback(state, configuration) {
...
 
+/**
+ * Call onAuthorizationFailed when 'FAILED' state is reached.
+ *
+ * @param {JSON} state          The current state
+ * @param {JSON} configuration  The redirectless configuration object
+ */
+export function failedStateCallback(state, configuration) {
+  if (state.status === 'FAILED') {
+    if(configuration.onAuthorizationFailed) {
```
[the remark this answers is not in the document]
Comment (mketchum) [reply src/utils/redirectless.js]: Good eyes 👀

File: types/index.d.ts (hunk 3 above)
Comment (mketchum) [reply types/index.d.ts]: whooops -- Fixed in https://github.com/pingidentity/pf-authn-js-widget/pull/100/commits/914a7d8b8d95530bbff0b9f859b70efbd9da37b2

**Review by khosravian — APPROVED** [review]

**Review by danielhoskin-ping — APPROVED** [review]

**Pull request merged**